src/checkers/inference/solver/backend/ExternalSolver.java

```diff
@@ -0,0 +1,161 @@
+package checkers.inference.solver.backend;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.util.Collection;
+
+import org.checkerframework.javacutil.BugInCF;
+
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Slot;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.util.FileUtils;
+import checkers.inference.solver.util.SolverEnvironment;
+
+/**
+ * Abstract solver which extends {@link Solver} with helper methods to invoke a custom external
+ * solver program.
+ *
+ * @param <T>
+ *            type of FormatTranslator required by this Solver
+ *
+ * @see Solver
+ */
+public abstract class ExternalSolver<T extends FormatTranslator<?, ?, ?>> extends Solver<T> {
+
+    public ExternalSolver(SolverEnvironment solverEnvironment, Collection<Slot> slots,
+            Collection<Constraint> constraints, T formatTranslator, Lattice lattice) {
+        super(solverEnvironment, slots, constraints, formatTranslator, lattice);
+    }
+
+    /**
+     * Functional interface for lambdas that process the stdOut or stdErr of the external solver
+     * process. Each lambda is given a {@link BufferedReader} as an input wrapping the stdOut or
+     * stdErr {@link InputStream}s.
+     */
+    @FunctionalInterface
+    protected interface StdOutputHandler {
```

Comment:
Use a better name as it is used for stderr and stdout.

Reply:
I think you might be able to use https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html
as `Consumer<BufferedReader>` without having to define your own functional interface.